src/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -1124,17 +974,329 @@
    "ALTER TABLE fact_paths DROP COLUMN value_type_id"
    "ALTER TABLE fact_values DROP COLUMN path_id"))
 
-(defn fill-in-null-producer-timestamp
-  "The producer_timestamp column can sometimes contain null values,
-  especially when updating from older versions. New use cases require
-  a value in that position, so we'll use the value from the
-  'timestamp' column if it's not there. "
+(defn version-2yz-to-300-migration
+  ;; This migration includes:
+  ;;   Insertion of the factsets hash column
+  ;;   Using a report_id (surrogate key) instead of the reports.hash (natural key)
+  ;;   Insert a noop column into reports
+  ;;   Drop latest_report table, merge with certnames
+  ;;   Change name to certname where inconsistent
+  ;;   Insert reports metrics and logs
+  ;;   Ensuring producer_timestamp is NOT NULL
+  ;;   Changing the types of hashes and uuids in postgres to bytea and uuid respectively
```

Comment:
Another option or middle ground here is support fast forward migrations. That is, we implement a system that keeps the discrete migrations while also allow us to provide a fast forward migration for major upgrades, that reduces the upgrade, but means we keep the history of updates (like in this case we could make a 29...33 migration). Has the added benefit of eventually allowing us to collapse all migrations to. I guess for us though, it means taking on more complication.